File: abc/136/e.py
from itertools import accumulate

N, K = map(int, input().split())
A = list(map(int, input().split()))


# Aの要素をK回以下の操作ですべてdの倍数にできるならば真を返す関数
def f(d):
    B = sorted(a % d for a in A)
    C = [d - b for b in B]
    X = list(accumulate(B))
    Y = list(accumulate(C))
    return any(
        X[j] == (Y[-1] - Y[j]) <= K
        for j in range(N)
    )


# 解の候補はsum(A)の約数である
S = sum(A)
ans = max(
    (
        S // d if f(S // d) else
        d if f(d) else
        0
    )
    for d in range(1, int(S**0.5) + 2)
    if S % d == 0
)

print(ans)

File: abc/136/e.rs
macro_rules! get {
    (Vec<$t:ty>) => {
        {
            let mut line: String = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            line.split_whitespace()
                .map(|t| t.parse::<$t>().unwrap())
                .collect::<Vec<_>>()
        }
    };
    ($t:ty) => {
        {
            let mut line: String = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            line.trim().parse::<$t>().unwrap()
        }
    };
    ($($t:ty),*) => {
        {
            let mut line: String = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            let mut iter = line.split_whitespace();
            (
                $(iter.next().unwrap().parse::<$t>().unwrap(),)*
            )
        }
    };
    ($t:ty; $n:expr) => {
        (0..$n).map(|_|
            get!($t)
        ).collect::<Vec<_>>()
    };
    ($($t:ty),*; $n:expr) => {
        (0..$n).map(|_|
            get!($($t),*)
        ).collect::<Vec<_>>()
    };
}

fn main() {
    let (N, K) = get!(usize, i64);
    let A = get!(Vec<i64>);

    // 解の候補はsum(A)の約数である
    let s = A.iter().sum::<i64>();
    let D = (1..s).take_while(|d| d * d <= s).collect::<Vec<_>>();
    let D2 = (1..s).take_while(|d| d * d <= s).map(|d| s / d).collect::<Vec<_>>();
    let ans = D.iter().chain(D2.iter())
        .filter(
            |&d|
            {
                let mut B = A.iter()
                    .map(|a| a % d)
                    .collect::<Vec<_>>();
                B.sort();
                let mut acc = vec!(0; N + 1);
                let mut rev_acc = vec!(0; N + 2);
                for i in 0..N {
                    acc[i + 1] = B[i] + acc[i];
                    rev_acc[N - i] = (d - B[N - i - 1]) + rev_acc[N - i + 1]
                }
                (0..(N + 1)).any(|i| acc[i] <= K && acc[i] == rev_acc[i + 1])
            }
        )
        .max()
        .unwrap();

    println!("{}", ans);
}